yeat/bandage/Snakefile

```diff
@@ -0,0 +1,64 @@
+# -------------------------------------------------------------------------------------------------
+# Copyright (c) 2022, DHS. This file is part of YEAT: http://github.com/bioforensics/yeat
+#
+# This software was prepared for the Department of Homeland Security (DHS) by the Battelle National
+# Biodefense Institute, LLC (BNBI) as part of contract HSHQDC-15-C-00064 to manage and operate the
+# National Biodefense Analysis and Countermeasures Center (NBACC), a Federally Funded Research and
+# Development Center.
+# -------------------------------------------------------------------------------------------------
+
+from glob import glob
+from pathlib import Path
+import re
+
+
+rule all:
+    input:
+        expand("analysis/bandage/{assembler}/.done", assembler=config["assemblers"])
+
+
+rule spades:
+    output:
+        "analysis/bandage/spades/.done"
+    input:
+        glob("analysis/spades/*.gfa"),
+        glob("analysis/spades/*.fastg")
+    run:
+        for gfa in input:
+            filename = Path(gfa).stem
+            shell("Bandage image {gfa} analysis/bandage/spades/{filename}.jpg")
+        shell("touch {output[0]}")
+
+
+def get_and_filter_contig_files():
+    pattern = r"analysis/megahit/intermediate_contigs/k\d+.contigs.fa"
+    contigs = glob(r"analysis/megahit/intermediate_contigs/k*.contigs.fa")
+    return filter(re.compile(pattern).match, contigs)
+
+
+rule megahit:
+    output:
+        "analysis/bandage/megahit/.done"
+    input:
+        get_and_filter_contig_files()
```

Comment:
As pointed out in the recent bug, to avoid hard-coded input file names, we are doing a glob search now in the rule `megahit` as well.

When applying this change, we have to ensure that the contig files that `megahit_toolkit` converts are the `k*.contigs.fa`--nothing more than that! In `analysis/megahit/intermediate_contigs`, megahit produces a lot of different`contigs.fa` files for each k-mer. For example, `k[blah].final.contigs.fa`. These "final" files cannot be converted into fastg files.

In consequence, I created a helper function called `get_and_filter_contigs_files()`.

In this function, I have a regex pattern and a glob function that retrieves all files matching with the Unix wildcard. I kept the glob retrieve function because it keeps the relative path of the return list of files and was the easiest to work with when doing the various shell commands.

Fun fact, glob only works with Unix wildcard input strings and not regex. In order to filter the list of glob'ed files, you need to do a regex filter.

I couldn't find another better way than this.

Reply:
Yeah, I'm not sure we'll get it much cleaner than that. Thanks!